### PR TITLE
Rescue exceptions from batch processing

### PR DESCRIPTION
**Pull request by FVolonnino**
Added a rescue for exceptions during batch processing and a new error type.

### CODE REVIEW

**Comment by FVolonnino** [comment]
> if a block is always required, you can check for its presence using `block_given?`
> 
> I don't like much this error handling, because it hides where the error is in your code, for example, trying to find why the error is a division by 0 in your example may be quite troublesome.
> 
> Do you have a better idea for this cases ?

**Comment by holywyvern** [comment]
Reopened as volo's pettition

**Comment by FVolonnino** [comment]
> Do you have a better idea for this cases ?
I could include the original exception and the first few lines in its backtrace to help with the debugging. What do you think?

**Comment by redwarewolf** [comment]
Adding to @holywyvern 's comment, I would recommend adding a backtrace to help find what is the error that caused the strategy to fail. Adding this exception should only emphasize on the fact that the strategy failed to the person who uses this gem, but it shouldn't hinder his work to actually fix the custom strategy that he develops.

Also, I'd recommend modifying the readme.md explaining and detailing this new functionality, explaining what this exception is and why it can arise when using this gem in the section for custom strategies. I can provide help with that task if needed. We will also be changing the version of the gem to 1.0.0 after merging this PR but that's something I'll do.

File: lib/wor/batchifier/exceptions.rb [hunk 1]
```diff
@@ -5,6 +5,7 @@ class WrongStrategy < StandardError; end
       class ExistingStrategy < StandardError; end
       class InterfaceNotImplemented < StandardError; end
       class StrategyNotFound < StandardError; end
+      class ProcessingError < StandardError; end
```
Comment (redwarewolf) [inline lib/wor/batchifier/exceptions.rb]: I'm still not %100 convinced with this name for the exception. This should be a total giveaway that the merge strategy defined on a custom strategy has failed for some reason, I'll keep thinking about the name for this exception.

File: spec/wor/batchifier/batchifier_spec.rb [hunk 2]
```diff
@@ -124,6 +124,19 @@
           expect(batchified_process).to eq(expected_response)
         end
       end
+
+      context 'when an error occurs while processing' do
```
Comment (redwarewolf) [inline spec/wor/batchifier/batchifier_spec.rb]: I would add the opposite case too, that it should not rise any exceptions on valid strategies.

File: lib/wor/batchifier/exceptions.rb (hunk 1 above)
Comment (redwarewolf) [reply lib/wor/batchifier/exceptions.rb]: Maybe something in the lines of "CustomStrategyMergingError"

Comment (redwarewolf) [reply lib/wor/batchifier/exceptions.rb]: @holywyvern what you think about this?

File: lib/wor/batchifier.rb [hunk 3]
```diff
@@ -12,16 +12,24 @@ module Batchifier
...
         merge(response, rec, strategy)
       end
     end
 
     private
 
+    def batch_process(chunk)
+      yield chunk
+    rescue StandardError => e
+      puts "Backtrace:\n\t#{e.backtrace.join("\n\t")}"
```
Comment (holywyvern) [inline lib/wor/batchifier.rb]: esto debería integrarse con un logger, preferiblemente

Comment (redwarewolf) [reply lib/wor/batchifier.rb]: That's a valid suggestion. What you think about something like this?

https://stackoverflow.com/questions/43348556/how-can-i-nicely-print-a-formatted-exception-that-i-catch-in-a-rescue-block/43348709

File: lib/wor/batchifier.rb [hunk 4]
```diff
@@ -12,16 +12,24 @@ module Batchifier
...
+      raise Wor::Batchifier::Exceptions::CustomStrategyMergingError,
+        "Custom strategy could not be merged"
```
Comment (redwarewolf) [inline lib/wor/batchifier.rb]: This isn't very accurate. I'd say that the custom merge strategy defined by the developer raised an exception during execution. Maybe we could rephrase it like:

"Custom Merge Strategy from #{ TheCustomStrategy.class.name } raised # { TheException.class.name } during execution.
Backtrace: ............. "

What you think about this?

Comment (FVolonnino) [reply lib/wor/batchifier.rb]: Ok, I will add this and the other suggested changes.

File: lib/wor/batchifier.rb [hunk 5]
```diff
@@ -12,16 +12,25 @@ module Batchifier
     # => optionaly a batch_size and a merge strategy. It will slice the collection and
     # => apply the chozen strategy to all chunks and merge the results. It expects the responses
     # => to be Hash. It can ignore them if the given strategy is no_response
-    def execute_in_batches(collection, batch_size: 100, strategy: :add)
+    def execute_in_batches(collection, batch_size: 100, strategy: :add, &block)
       strategy_class = classify_strategy(strategy)
       collection.lazy.each_slice(batch_size).inject(strategy_class.new.base_case) do |rec, chunk|
-        response = yield(chunk)
+        response = batch_process(chunk, strategy, &block) if block_given?
```
Comment (redwarewolf) [inline lib/wor/batchifier.rb]: What if there is are no blocks given for processing the batch? This escapes the scope of this PR but I'll add this as an issue to raise an exception on incorrect usage of the gem.

**Comment by redwarewolf** [comment]
After giving a quick look to this PR it occurred to me that perhaps we could allow user of the gem to give a block of code to the `execute in batches()`

Something along the lines of:

`  execute_in_batches(information, batch_size: 100, `
    `                                  strategy: { |response, rec| response.merge(rec) { |_, v1, _| v1 } }) do |chunks| `
  `    some_process(chunks) `
`  end `

It's something I'll definitely look into that seems it could be useful as you do not need to always define a very complex merging strategy.

**Pull request closed without merging**